### ISSUE TITLE
Adding a pattern adds two patterns, additionally one at the end

### ISSUE THREAD

**Issue by michaelsjackson** [JoepVanlier/Hackey-Patterns#1]
Is the second added pattern needed when adding a new pattern in Reaper arrange?

**Comment by JoepVanlier** [comment]
Yes. It's to make sure you still have the pattern prototype if you delete it from the playlist.

It's a bit early for issues, since I don't even have a working proto yet. I'll make a forum post when it's ready for actual testing ;)

**Comment by michaelsjackson** [comment]
As an additional side idea, pattern prototypes, could be saved simply as external .mid files on harddisk, in a subdir in project directory called 'pattern pool'. Imo this would be most elegant solution, not the fastest maybe, but most powerful.

**Comment by JoepVanlier** [comment]
No. There are several reasons why in the API this would be a massive headache.

- This would require constant writing and reading of MIDI files every time a pattern in the pool is edited.
- It would also require constant checking whether the MIDI data has changed. And it would require all MIDI items to be polluted with a reference to their filename (which has to be unique, so pattern name is not an option since the user can edit this). GUID is also not an option here, since that's different every time the MIDI item would be imported.
- Every time a track is added, the MIDI files would have to be reorganized and it would still not eliminate a need for the MIDI item to be stored somewhere in the project since there has to be a pool reference item to get the length of the MIDI item in the current project. MIDI items live on their own timescale.

**Comment by michaelsjackson** [comment]
No, midi files can be used as linked .mid files, then all .mid files are automatically like pooled copies. You never need to sync anything as you have always only one variant, the one which is on harddisk. I am using .mid files mostly in this form, within Reaper. Setting all items offline, then back online, does the updating, in case you changed the files on harddisk, even using other/external midi editing tools. Reaper updates immediately, very nice! To me this method is most elegant.

**Comment by michaelsjackson** [comment]
Ok, but at least, it seems that method could work for notes and midi cc. One day, an experimental release could be tried out maybe... just another idea for the future, what could be tested. I like usually solutions which have:
- simplest solutions
- the more independence the better (here: independence of arrange)
- separate .mid files, if possible (as in this case, I learned this in my 25 years midi experience life)

**Comment by JoepVanlier** [comment]
Two issues would need to be resolved first. Possibly in a future REAPER release:

- For MIDI storage, we have SWS command 40685 (convert active take MIDI to .mid file reference). Problem is, we have no reliable way of polling what filename this midi reference will get. Nor can we store what track it is meant for without storing custom information in them or the project file (neither solution is very appealing) and then parsing them. I also don't see an API for reading them and reading their properties.

- For automation storage we don't have API calls that can export them as far as I can see. You can save them from the REAPER arrange view, but I don't see what command ID that is using (if any). I also don't see a way to import them.

For now I am shelving this in favor of getting a working sequencer first. Feel free to try and experiment with the code to get it to work the way you want though.

**Comment by michaelsjackson** [comment]
First getting a working sequencer, mostly for drum editing, is a great idea. I would use it mostly for drums anyway.
